### PR TITLE
Add masking to sliceviewer

### LINKED ISSUE #39908: Add ability to mask workspaces using sliceviewer
Add ability to mask workspaces using sliceviewer

### PR DESCRIPTION
**Pull request by MialLewis**
### Description of work

Adds the ability to generate mask table workspaces and apply masks using sliceviewer.

Closes [#39908](https://github.com/mantidproject/mantid/issues/39908)

### To test:

1) Run the following script:
```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

nrows = 100
nbins = 100
output_ws = WorkspaceFactory.create("Workspace2D", NVectors=nrows,
                                    XLength=nbins, YLength=nbins)

create_histo_data = True
x_data = [x for x in range(nbins)]
xdata = np.array(x_data) # filled with 0->9
ydata = [y for y in range(nrows)]
edata = np.ndarray(shape=(nbins))
edata.fill(0)

for i in range(nrows):
    output_ws.setX(i, xdata)
    output_ws.setY(i, ydata + np.ones(nrows)*i)
    output_ws.setE(i, edata)
if create_histo_data:
    ConvertToHistogram(output_ws, OutputWorkspace="out_ws")
else:
    mtd.addOrReplace("out_ws", output_ws)
```

2) Open sliceviewer on the workspace, ensure colorbar is linear scaling.

3) Click the `Toggle masking on/off` icon on the toolbar.

4) Draw a selection of masks on the sliceviewer plot using the rectangle, ellipse and polygon tools.

5) Once drawn, click `export drawn mask to table`. Sense check resultant table workspace.

6) Click `Apply drawn mask`. Is mask drawn correctly.

7) Stress test the feature. Does the behaviour of the selectors make sense?

---

### Reviewer

**Your comments will be used as part of the gatekeeper process.** Comment clearly on what you have checked and tested during your review. Provide an audit trail for any changes requested.

As per the [review guidelines](http://developer.mantidproject.org/ReviewingAPullRequest.html):

- Is the code of an acceptable quality? ([Code standards](http://developer.mantidproject.org/Standards/)/[GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html))
- Has a thorough functional test been performed? Do the changes handle unexpected input/situations?
- Are appropriately scoped unit and/or system tests provided?
- Do the release notes conform to the [guidelines](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html) and describe the changes appropriately?
- Has the relevant (user and developer) documentation been added/updated?

### Gatekeeper

As per the [gatekeeping guidelines](https://developer.mantidproject.org/Gatekeeping.html):

- Has a thorough first line review been conducted, including functional testing?
- At a high-level, is the code quality sufficient?
- Are the base, milestone and labels correct?

### CODE REVIEW

**Comment by MialLewis** [comment]
@coderabbitai full review

File: qt/python/mantidqt/mantidqt/widgets/sliceviewer/models/masking.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+
+    def _generate_line(self, start, end):
+        y_index = 1 if not self._transpose else 0
+        x_index = 0 if not self._transpose else 1
+        start_y = self.snap_to_bin_centre(start[y_index])
+        end_y = self.snap_to_bin_centre(end[y_index])
+        start_x = start[x_index]
+        end_x = end[x_index]
+        m = (start_y - end_y) / (start_x - end_x)
+        c = start_y - m * start_x
```
[the remark this answers is not in the document]
Comment (MialLewis) [reply qt/python/mantidqt/mantidqt/widgets/sliceviewer/models/masking.py:200]: A division by 0 results in an `inf` value, this is handled elsewhere. In `_calculate_relevant_x_value_pairs` of the `masking` model.

File: qt/python/mantidqt/mantidqt/widgets/sliceviewer/presenters/selector.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+
+def make_selector_class(base, clear_handle=None):
+    def in_axes_event(self, event):
+        """
+        Only process event if inside the axes with which the selector was init
+        This fixes bug where the x/y of the event originated from the line plot axes not the colorfill axes
+        """
+        return event.inaxes is None or self.ax == event.inaxes.axes
+
```
[the remark this answers is not in the document]
Comment (MialLewis) [reply qt/python/mantidqt/mantidqt/widgets/sliceviewer/presenters/selector.py:54]: Not changing this, existing working code.

File: qt/python/mantidqt/mantidqt/widgets/sliceviewer/views/dataview.py [hunk 3]
```diff
@@ -482,18 +502,54 @@ def on_data_limits_changed(self):
...
+        React to export masking selected
+        """
+        self.presenter.export_masking_clicked()
+
+    def on_apply_masking_clicked(self):
+        """
+        React to apply masking selected
+        """
+        self.presenter.apply_masking_clicked()
+
```
[the remark this answers is not in the document]
Comment (MialLewis) [reply qt/python/mantidqt/mantidqt/widgets/sliceviewer/views/dataview.py:548]: These methods are abstract on the presenter base class, which enforces them.

File: qt/python/mantidqt/mantidqt/widgets/sliceviewer/models/masking.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+
+    def apply_selectors(self):
+        mask_ws = self.generate_mask_table_ws(store_in_ads=False)
+        alg = AlgorithmManager.create("MaskBinsFromTable")
+        alg.initialize()
+        alg.setChild(True)
+        alg.setProperty("InputWorkspace", self._ws_name)
+        alg.setProperty("OutputWorkspace", self._ws_name)
+        alg.setProperty("MaskingInformation", mask_ws)
+        alg.execute()
```
[the remark this answers is not in the document]
Comment (MialLewis) [reply qt/python/mantidqt/mantidqt/widgets/sliceviewer/models/masking.py:295]: This is not a pattern used in mantid, if there is a failure when executing the `MaskBinsFromTable` algorithm, that algorithm will throw.

File: qt/python/mantidqt/mantidqt/widgets/sliceviewer/presenters/masking.py [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+    def set_active(self, state):
+        self._selector.set_active(state)
+
+    def clear(self):
+        if self._clear:
+            return
+        for artist in self._selector.artists:
+            artist.remove()
+        self._clear = True
+
```
[the remark this answers is not in the document]
Comment (MialLewis) [reply qt/python/mantidqt/mantidqt/widgets/sliceviewer/presenters/masking.py:38]: All classes that inherit from this ABC set `_selector`. This persists for the lifetime of the object.

There also doesn't appear to be any issue with the order of disconnect and artist removal.

File: qt/python/mantidqt/mantidqt/widgets/sliceviewer/presenters/masking.py [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+        :param eclick: Event marking where the mouse was clicked
+        :param erelease: Event marking where the mouse was released
+        """
+        cinfo_click = cursor_info(self._img, eclick.xdata, eclick.ydata)
+        if cinfo_click is None:
+            return
+        cinfo_release = cursor_info(self._img, erelease.xdata, erelease.ydata)
+        if cinfo_release is None:
+            return
+
```
[the remark this answers is not in the document]
Comment (MialLewis) [reply qt/python/mantidqt/mantidqt/widgets/sliceviewer/presenters/masking.py:96]: This is an existing pattern that works well, taken from `qt/python/mantidqt/mantidqt/widgets/sliceviewer/presenters/lineplots.py`, I won't be changing anything.

File: qt/python/mantidqt/mantidqt/widgets/sliceviewer/presenters/presenter.py [hunk 7]
```diff
@@ -214,6 +217,11 @@ def dimensions_changed(self):
             else:
                 data_view.disable_tool_button(ToolItemText.NONORTHOGONAL_AXES)
 
+        # Reset masking if dimensions changed
+        if self.view.data_view.masking:
+            self.view.data_view.masking.clear_and_disconnect()
+            self.view.data_view.masking.clear_model()
+
```
[the remark this answers is not in the document]
Comment (MialLewis) [reply qt/python/mantidqt/mantidqt/widgets/sliceviewer/presenters/presenter.py:224]: This is not an issue `self.view.data_view.masking.clear_and_disconnect()` cleans the selectors.

**Comment by MialLewis** [comment]
@coderabbitai remove all of your comments

**Review by MohamedAlmaki — COMMENTED** [review]
Is this nan value expected in this case?

Image

**Comment by MialLewis** [comment]
> Is this nan value expected in this case?
> 
> Image

Were you able to replicate this, I'm struggling.

File: qt/python/mantidqt/mantidqt/widgets/sliceviewer/presenters/selector.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        trans = BboxTransform(boxin=full_bbox, boxout=array_extent)
+    point = trans.transform_point([ydata, xdata])
+    if any(np.isnan(point)):
+        return None
+
+    point = point.astype(int)
+    if 0 <= point[0] <= arr.shape[0] and 0 <= point[1] <= arr.shape[1]:
+        return CursorInfo(array=arr, extent=extent, point=point, data=(xdata, ydata))
+    else:
+        return None
```
[the remark this answers is not in the document]
Comment (MialLewis) [reply qt/python/mantidqt/mantidqt/widgets/sliceviewer/presenters/selector.py:44]: This suggestion introduced a bug.

**Review by MohamedAlmaki — COMMENTED** [review]
I have tested the functionality, and it works as expected. The simplification resolved the issue of the None values. It no longer happens or could happen, but less frequently. We agreed to leave them as these edge cases are rare. Happy to approve once the tests pass.

**Review by MohamedAlmaki — APPROVED** [review]

**Comment by SilkeSchomann** [comment]
@MialLewis Please open a separate PR with the release note.
